RFC-0011-coordinator-throttling.md

```diff
@@ -0,0 +1,154 @@
+## Coordinator throttling mechanism
+
+Proposers
+
+* Prashant Golash (@PRASHANT GOLASH)
+
+
+## Summary
+
+Implement a mechanism to prevent queries from being admitted when a cluster is experiencing overload. 
+A cluster is considered overloaded based on aggregated metrics reported by its underlying worker nodes. 
+
+These metrics can include CPU utilization, memory usage, and task queue lengths. By analyzing this aggregated information, the system can make informed decisions about the cluster’s current load status.
+The exact criteria used to determine overload are governed by customizable and extensible policies. For example, a policy might specify that a cluster is overloaded if a certain percentage or number of worker nodes reach critical resource thresholds (“run hot”).
+In our case, most of the workers get overloaded within a matter of few mins
+
+A sample graph of clusters running hot. All the workers are running hot in this case.
+
+![Design diagram](RFC-0011/cpuload.png)
+
+The impact is:
+
+- Large variance in execution time
+- Memory kills / Overload on dependency services like metastore / warm-storage etc 
+
+This provide more reasons to work on protecting cluster overload.
+
+## Background
+
+While not exhaustive, current throttling/possible mechanisms relevant for this discussion
+
+#### Granular task based scheduling
+
+**Leaf Task scheduling**
+
+The node selection logic in Presto involves randomly selecting 10 nodes from those that do not currently have scheduled splits for the given query. Among these 10 nodes, the system then chooses the node that is the least busy. For reference, the relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java#L182).
+Additionally, in the process of leaf task scheduling, a feedback mechanism is employed whereby splits are assigned to the nodes deemed least busy. It is important to note that in this context, "busyness" is defined by the number of splits being processed by a node, rather than its actual system load. 
+
+While this approach is generally effective for distributing work, it may not always result in the most optimal allocation of resources.
```

Comment:
Can you elaborate on the difference between "effectiveness when distributing work" vs. "allocating resources"?  To me they sound the same--resources are worker resources, so effective distribution of work means effective allocation of resources.

Reply:
In our scenario even with this logic of choosing least busy nodes based on splitting, most of the workers could become overload. I didn't meant to say this logic is wrong. Rather over-admitting is the key issue here.